### PR TITLE
Take advantage of Java8 language features. Misc cleanup.

### PR DESCRIPTION
**Pull request by aslakhellesoy**
## Summary

Use Java8 language features

## Details

We've recently upgraded to Java8 (by bumping `cucumber-parent`)

## Motivation and Context

The code is more terse

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [x] The change has been ported to Java.
- [ ] The change has been ported to Ruby.
- [ ] The change has been ported to JavaScript.
- [ ] The change has been ported to Go.
- [ ] The change has been ported to .NET.
- [ ] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: cucumber-expressions/java/pom.xml [hunk 1]
```diff
@@ -36,7 +36,7 @@
         <dependency>
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
-            <version>4.12</version>
+            <version>4.13-beta-3</version>
```
Comment (mpkorstanje) [inline cucumber-expressions/java/pom.xml]: We could use JUnit Jupiter instead. I think that is preferable to using a beta version.

File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/TreeRegexp.java [hunk 2]
```diff
@@ -74,38 +75,10 @@ Pattern pattern() {
...
         if (!matcher.matches()) return null;
-        return groupBuilder.build(matcher, new IntRange(0, matcher.groupCount() + 1));
+        return groupBuilder.build(matcher, IntStream.rangeClosed(0, matcher.groupCount()).iterator());
     }
 
     public GroupBuilder getGroupBuilder() {
         return groupBuilder;
     }
-
-    private static class IntRange implements Iterator<Integer> {
```
Comment (mpkorstanje) [inline cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/TreeRegexp.java]: Good riddance.

File: cucumber-expressions/java/pom.xml (hunk 1 above)
Comment (aslakhellesoy) [reply cucumber-expressions/java/pom.xml]: Eh JUnit Jupiter - is that a new API I have to learn?

Comment (aslakhellesoy) [reply cucumber-expressions/java/pom.xml]: I don't think there is much to risk using a beta version for a test dependency

Comment (mpkorstanje) [reply cucumber-expressions/java/pom.xml]: > Eh JUnit Jupiter - is that a new API I have to learn?

Also known as JUnit 5. For the sort of tests we are using, you'd have to use some new imports, that is it.

I'll update it to Junit 5 on the next pass through. It's not so difficult.

**Pull request merged**